src/ros2cs/ros2cs_core/Ros2cs.cs

```diff
@@ -208,28 +213,46 @@ public static void SpinOnce(List<INode> nodes, double timeoutSec = 0.1)
 
         // TODO - This can be optimized so that we cache the list and invalidate only with changes
         var allSubscriptions = new List<ISubscriptionBase>();
+        var allClients = new List<IClientBase>();
+        var allServices = new List<IServiceBase>();
         foreach (INode node_interface in nodes)
         {
           Node node = node_interface as Node;
           if (node == null)
             continue; //Rare situation in which we are disposing
-
-          foreach(ISubscriptionBase subscription in node.Subscriptions)
-          {
-            if (subscription == null)
-              continue; //Rare situation in which we are disposing
-
-            allSubscriptions.Add(subscription);
-          }
+          
+          allSubscriptions.AddRange(node.Subscriptions.Where(s => s != null));
+          allClients.AddRange(node.Clients.Where(c => c != null));
+          allServices.AddRange(node.Services.Where(s => s != null));
         }
 
         // TODO - investigate performance impact
-        WaitSet.Wait(global_context, allSubscriptions, timeoutSec);
-
-        // Sequential processing
-        foreach (var subscription in allSubscriptions)
+        WaitSet.Resize(
+          (ulong)allSubscriptions.Count,
+          (ulong)allClients.Count,
+          (ulong)allServices.Count      
+        );
+        foreach(var subscription in allSubscriptions)
+        {
+          AddResult result = WaitSet.TryAddSubscription(subscription, out ulong _);
+          Debug.Assert(result != AddResult.FULL, "no space for subscription in WaitSet");
+        }
+        foreach(var client in allClients)
+        {
+          AddResult result = WaitSet.TryAddClient(client, out ulong _);
+          Debug.Assert(result != AddResult.FULL, "no space for client in WaitSet");
+        }
+        foreach(var service in allServices)
+        {
+          AddResult result = WaitSet.TryAddService(service, out ulong _);
+          Debug.Assert(result != AddResult.FULL, "no space for Service in WaitSet");
+        }
+        if (WaitSet.Wait(TimeSpan.FromSeconds(timeoutSec)))
```

Comment:
This is documented for [rcl_wait](http://docs.ros.org/en/humble/p/rcl/generated/function_wait_8h_1a732278988c802fe2c9d8ec24752f9dd9.html#exhale-function-wait-8h-1a732278988c802fe2c9d8ec24752f9dd9) and I forgot to handle it.
I would solve this by calling `Thread.Sleep` to sleep until the timeout expires if we have nothing to wait on since we check for new subscriptions, clients and services every spin and sleeping would be the same as the wait set timing out.
One alternative would be to return immediately but this could lead to excessive spinning until the first "waitable" is added.
To support both use cases `SpinOnce` could return a bool indicating if spinning was successful and the spin loop could sleep if not.

Reply:
We could also try to handle the `RCLReturnEnum.RCL_RET_WAIT_SET_EMPTY` ret code in `WaitSet.cs` and `return false` in `Wait()` if we get it.

This should do the trick as well :thinking:

Reply:
I think an exception would be better since we have to differentiate between a timeout and an empty wait set.